apps/avatax/src/modules/saleor/order.ts

```diff
@@ -47,3 +51,51 @@ export class SaleorOrder {
     return this.data.order.channel.taxConfiguration.pricesEnteredWithTax;
   }
 }
+export class SaleorCancelledOrder {
+  private static schema = z.object({
+    order: z.object({
+      channel: z.object({
+        id: z.string(),
+        slug: z.string(),
+      }),
+      id: z.string(),
+      avataxId: z.string(),
+    }),
+    recipient: z.object({
+      privateMetadata: z.array(
+        z.object({
+          key: z.string(),
+          value: z.string(),
+        }),
+      ),
+    }),
+    __typename: z.literal("OrderCancelled"),
+  });
+
+  private data: z.infer<typeof SaleorCancelledOrder.schema>;
+
+  constructor(data: OrderCancelledPayload) {
+    const verifiedPayload = verifyOrderCanceledPayload(data);
+
+    if (verifiedPayload.isErr()) {
+      throw verifiedPayload.error;
+    }
+
+    const ParsingError = BaseError.subclass("AvataxAppSaleorOrderCancelledParsingError");
+
+    const parse = Result.fromThrowable(SaleorCancelledOrder.schema.parse, ParsingError.normalize);
+
+    const parsedData = parse(data);
+
+    if (parsedData.isErr()) {
+      throw parsedData.error;
+    }
+
+    // Payload is  parsed and verified to be compatible with the schema
+    this.data = data as z.infer<typeof SaleorCancelledOrder.schema>;
+  }
+
+  public get payload() {
```

Comment:
**Question:** maybe we should get only `avataxId` here instead of full payload

Reply:
We should avoid exposing full payload. If we expose payload, we will perform operations on raw payload.

Instead, we should perform these operations inside the class, making exact schema syntax "private". Its also how we do DRY → single place to perform business operations